index.html

```diff
@@ -418,6 +419,58 @@ <h2>Transferable MediaStreamTrack</h2>
       <p>The underlying source is supposed to be kept alive between the transfer and transfer-receiving steps, or as long as the data holder is alive.
       In a sense, between these steps, the data holder is attached to the underlying source as if it was a track.</p>
     </div>
+    <h3>MediaStreamTrack frame counters</h3>
+    <div>
+      <p>
+        Frame counters allow you to tell why the frame rate experienced is
+        different from the expected one - for instance, the camera may be
+        slowing down frame production because it's dark, the settings applied
+        may produce decimation of frames, or the system may be CPU starved and
+        therefore dropping frames.
+      </p>
+    </div>
+    <div>
+      <p>The WebIDL changes to add frame counters are the following:
+      <pre class="idl"
+>partial interface MediaStreamTrack {
+  readonly attribute unsigned long long framesEmitted;
+  readonly attribute unsigned long long framesDropped;
+};</pre>
+    </div>
+    <section>
+      <h4>Attributes</h4>
+      <dl data-link-for="MediaStreamTrack" data-dfn-for="MediaStreamTrack"
+      class="attributes">
+        <dt>
+          <dfn>framesEmitted</dfn> of type {{unsigned long long}}, readonly
+        </dt>
+        <dd>
+          <p>
+            Only defined for video tracks. The total number of frames emitted by
+            the track's source for consumption by one or more sinks. See also
+            {{framesDropped}}.
+          </p>
+        </dd>
+        <dt>
+          <dfn>framesDropped</dfn> of type {{unsigned long long}}, readonly
+        </dt>
+        <dd>
+          <p>
+            Only defined for video tracks. Incremented every time a source frame
+            was dropped prior to being emitted. This happens when a frame is
+            decimated in order to achieve the desired
+            {{MediaTrackSettings/frameRate}} or if the user agent dropped the
+            frame for other reasons, such as if the pipeline is not consuming
+            emitted frames fast enough or if the track does not have any sinks.
```

Comment:
This doesn't sound like an inherent property of the track anymore.

The idea that frame-rate might depend on consumption seems problematic and potentially user agent dependent. I don't think this has come up before. If the actual frame-rate is determined downstream and depends on downstream, perhaps it is best measured downstream as well? See https://github.com/w3c/mediacapture-extensions/issues/76#issuecomment-1276459896.

cc @Pehrsons @padenot thoughts?

Reply:
It might make more sense to count a frame that survives until emission as a framesEmitted even if there are no sinks here, since it's the would-be FPS that any consumer would experience if there were any consumers.

No strong opinion but happy to change this

Reply:
I mostly wonder what the definition of "emission" and "pipeline" is.

But I get the gist of this: `framesDropped` would for instance, for a camera that only captures 30fps but with a track setting of 15fps, record those extraneous frames that had to be dropped to reach 15. Or it could record those frames dropped in order to not lag behind under high load, in case frames from a camera go through some processing/conversion step.

I was also thinking `framesDropped` would get pretty obvious with `VideoTrackGenerator`, but the processing happens in `MediaStreamTrackProcessor` and so does not apply. A processor could provide its own measure of frames dropped from its buffer, I guess.

What if a track is disabled? I imagine `framesDropped` would be incrementing at this point in case the source is still capturing? Might need clarification.

As long as `framesDropped` doesn't depend on consumption by track sinks, and to my understanding it doesn't, this seems fine to me.

Reply:
I'll create a slide for this for the next WG meeting on Tuesday, but I'm thinking of frames "before" or "after" reaching the media source. A frame produced that doesn't reach the source is "dropped" and a frame produced that does gets "emitted" to the rest of the pipeline. WebRTC for example is later in the pipeline and already has frame counters, but it doesn't know about frames dropped earlier and as such we can't tell the FPS of the camera (unless we assume 0 frame drops)

Reply:
What does "reach the source" mean?

Reply:
That the frame is ready to be handed off to sinks, if there are any.

We usually think of the track's source as the thing producing the frames, but in order for a frame to go from the camera to something consuming frames there could be several steps along the way.

Examples:
- The camera process receives a frame from the camera (++framesProduced), the frame is passed to the renderer process and eventually ends up where the media source forwards it to its sinks (++framesEmitted).
- The camera process receives a frame from the camera (++framesProduced), the frame is bassed to the renderer process but the renderer process' buffer pool is out of frames so the frame is discarded (++framesDropped).